### PR TITLE
Cpi 155 klarna

### PR DESCRIPTION
**Pull request by andrii-kovalenko-ct**
@butenkor @ahalberkamp @heshamMassoud @LEQADA 

In this PR we implemented Klarna payment support.

The major changes are:

  - [`BaseKlarnaRequest`](https://github.com/commercetools/commercetools-payone-integration/pull/159/files#diff-1de6827c78254378f726a5c9d7feb366), where mapping of line items to the request is implemented
  - [`KlarnaRequestFactory`](https://github.com/commercetools/commercetools-payone-integration/pull/159/files#diff-cab310e17385dc0e656d2d7debbf0b94), where Klarna specific custom fields are mapped
  - [`PayoneKlarnaCountryToLanguageMapper`](https://github.com/commercetools/commercetools-payone-integration/pull/159/files#diff-4d37df1c508ee0a9f2a770dea123881a), where country mapped to language, because Klarna payments are restricted to have specific country-language dependency

### CODE REVIEW

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java [hunk 1]
```diff
@@ -12,16 +12,23 @@
...
 import org.slf4j.LoggerFactory;
 
+import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.time.LocalDate;
 import java.time.format.DateTimeFormatter;
-import java.util.Locale;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
```
Comment (butenkor) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java]: To prevent class collisions, improve reading and compile time i would try to avoid usage of wildcards.

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java [hunk 2]
```diff
@@ -66,48 +78,39 @@ public static void mapBillingAddressToRequest(
...
-                    id = id.replace("-", "").substring(0, 20);
-                    request.setCustomerid(id);
-                } else {
-                    request.setCustomerid(customerNumber);
-                }
-            });
+        ofNullable(customer.getCustomerNumber())
+                .ifPresent(customerNumber -> {
+                    if (customerNumber.length() > 20) {
+                        LOG.warn("customer.customerNumber exceeds the maximum length of 20! Using substring of customer.id as fallback.");
```
Comment (butenkor) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java]: This should be mentioned in readme documentation.

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/klarna/KlarnaRequestFactory.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+     */
+    @Override
+    public KlarnaAuthorizationRequest createAuthorizationRequest(final PaymentWithCartLike paymentWithCartLike) {
+        return createRequestInternal(paymentWithCartLike, KlarnaAuthorizationRequest::new);
+    }
+
+    protected <BKR extends BaseKlarnaRequest> BKR createRequestInternal(final PaymentWithCartLike paymentWithCartLike,
+                                                                        final TriFunction<PayoneConfig, String, PaymentWithCartLike, BKR> requestConstructor) {
+        final Payment ctPayment = paymentWithCartLike.getPayment();
+        Preconditions.checkArgument(ctPayment.getCustom() != null, "Missing custom fields on payment!");
```
Comment (butenkor) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/klarna/KlarnaRequestFactory.java]: Payment ID would be useful in logs?

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/KlarnaAuthorizationRequest.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import com.commercetools.pspadapter.payone.domain.ctp.PaymentWithCartLike;
+import com.commercetools.pspadapter.payone.domain.payone.model.common.RequestType;
+
+/**
+ * <b>Caveat:</b> according to Payone discussion this request should not be used by the shops, because in this case
+ * <i>capture</i> state is activated automatically.
+ * Preferable way is to make <i>{@link KlarnaPreauthorizationRequest preauthorization}</i> request and then
+ * make direct <i>capture</i>, when the goods are packaged, shipped or delivered.
+ */
+public class KlarnaAuthorizationRequest extends BaseKlarnaRequest {
```
Comment (butenkor) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/KlarnaAuthorizationRequest.java]: Looking at comment i wonder why we need this class if the flow is not even recommended by Payone?

File: docs/Field-Mapping.md [hunk 5]
```diff
@@ -283,22 +283,30 @@ All payment methods:
     * `clearing_bankbic` ->  `paidFromBIC` 
    
 `INVOICE-KLARNA`:
-  * _Required_ `language` -> custom field `languageTag` of Type String on the CT Payment
-    * (may be moved to the Order, Cart and Customer Objects later)
+  * _Required_ 
+    * `language` -> custom field `languageTag` of Type String on the CT Payment
```
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: `language` is Payone request field, `languageTag` is legacy custom field in CTP project (guess descended from some old projects).

File: docs/Field-Mapping.md [hunk 6]
```diff
@@ -283,22 +283,30 @@ All payment methods:
...
+      (may be moved to the Order, Cart and Customer Objects later)
+      
+    * `birthday` -> String custom field `birthday` in Payment in _YYYMMDD_ format.
+    * `ip` -> the IP address of the user -> String custom field `ip` in Payment. 
+    * `gender` -> The value is suppressed to first lowercase character -> String custom field `gender` in Payment
+    * `personalid` -> (***Not supported yet***) Personal ID Nr. 
+          Mandatory for Klarna if customers billing address is in certain nordics countries.
+          -> `personalId` custom field of type String.
```
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: To remember about this mandatory field when we want to implement the service support in nordic countries.

File: docs/Project-Lifecycle.md [hunk 7]
```diff
@@ -324,6 +325,20 @@ Then you are able to review the tests results in [Test results page](http://comm
...
+### Klarna Testing notes
+
+Klarna has very strict requirements to the payment details. For testing purpose one should use
+[special Klarna test credentials](https://developers.klarna.com/en/de/kpm/test-credentials).
+
+These tests require next CTP settings:
+
+  * `gender`, `ip` and `birthday` custom fields in _payment-INVOICE-KLARNA_ custom type
+  * German language, country and prices must be active.
+  * 19% German tax is active
```
Comment (andrii-kovalenko-ct) [inline docs/Project-Lifecycle.md]: I mean this:
screen shot 2017-07-11 at 18 49 38

File: docs/Project-Lifecycle.md [hunk 8]
```diff
@@ -324,6 +325,20 @@ Then you are able to review the tests results in [Test results page](http://comm
...
+  * `test-999-cent-code` and `test-10-percent-code` discount codes and respective discounts are active
+  * products from test mock cart [KlarnaCartWithTestAccountAddress.json](/blob/master/functionaltests/src/test/resources/mocks/paymentmethods/klarna/https://github.com/commercetools/commercetools-payone-integration/blob/17da0f6ed1c4b4b1e0b6d561fe03d8cfa3c0dc38/functionaltests/src/test/resources/mocks/paymentmethods/klarna/KlarnaCartWithTestAccountAddress.json)
+   are published
```
Comment (andrii-kovalenko-ct) [inline docs/Project-Lifecycle.md]: requires separate issue/task/time

File: service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java [hunk 9]
```diff
@@ -39,16 +39,17 @@ public PaymentWithCartLike dispatchPayment(PaymentWithCartLike paymentWithCartLi
...
-            .orElseThrow(() -> new IllegalArgumentException(String.format(
-                    "Unsupported payment method '%s'", paymentMethodInfo.getMethod())));
+        return Optional.of(PaymentMethod.fromMethodKey(paymentMethodInfo.getMethod()))
+                .map(methodDispatcher::get)
+                .map(dispatcher -> {
+                    try {
+                        return dispatcher.dispatchPayment(paymentWithCartLike);
+                    } catch (final ConcurrentModificationException cme) {
+                        throw new java.util.ConcurrentModificationException("The payment could not be dispatched: "
+                                + cme.getMessage(), cme);
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java]: legacy code - can't explain, unfortunately.

File: service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java [hunk 10]
```diff
@@ -53,14 +53,14 @@ public PaymentHandleResult handlePayment(@Nonnull final String paymentId) {
         try {
             for (int i = 0; i < RETRIES_LIMIT; i++) {
                 try {
-                    final PaymentWithCartLike payment = commercetoolsQueryExecutor.getPaymentWithCartLike(paymentId);
-                    String paymentInterface = payment.getPayment().getPaymentMethodInfo().getPaymentInterface();
+                    final PaymentWithCartLike paymentWithCartLike = commercetoolsQueryExecutor.getPaymentWithCartLike(paymentId);
+                    String paymentInterface = paymentWithCartLike.getPayment().getPaymentMethodInfo().getPaymentInterface();
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java]: code prettify-formatting - requires separate issue/task/branch

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/PayonePostServiceImpl.java [hunk 11]
```diff
@@ -58,6 +68,43 @@ public static PayonePostServiceImpl of(final String payoneServerApiUrl) throws I
...
+     * Example:
+     * <pre> {"a": [5, 7], "b": "c", "d": ["e", "f"]} =>
+     * {"a[1]": 5, "a[2]": 7, "b": "c", "d[1]": "e", "d[2]": "f"}
+     * </pre>
+     *
+     * @param parameters Set of key-value pairs to expand
+     * @return new {@link Map} where keys with list values
+     */
+    @Nonnull
+    Map<String, Object> getObjectMapWithExpandedLists(Map<String, Object> parameters) {
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/PayonePostServiceImpl.java]: because these field names are dynamic: list with name `a` and values `[5,6]` will be converted to request parameters `a[1]=5&a[2]=6`.

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/BaseKlarnaRequest.java [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+        this.id = new ArrayList<>(itemsCount);
+        this.it = new ArrayList<>(itemsCount);
+        this.pr = new ArrayList<>(itemsCount);
+        this.no = new ArrayList<>(itemsCount);
+        this.de = new ArrayList<>(itemsCount);
+        this.va = new ArrayList<>(itemsCount);
+
+        // accumulator for full prices - without discounts
+        final AtomicLong accumulatedFullPriceCents = new AtomicLong(0L);
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/BaseKlarnaRequest.java]: the reason is different: the closure below (in `forEach` requires  references to be immutable instances, that's why use this _wrapper_: to count inside closures.

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java (hunk 1 above)
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java]: It's common pattern to squeeze imports to wildcard if there are more than N imports of the same package. The same opinion in [Clean Code](https://stackoverflow.com/a/1992695/907576). The same opinion in default _Intellij Idea_ settings, the same opinion i have :)

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java (hunk 2 above)
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java]: it is: https://github.com/commercetools/commercetools-payone-integration/blob/master/docs/Field-Mapping.md#commercetools-payment-resource-mapping

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/klarna/KlarnaRequestFactory.java (hunk 3 above)
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/klarna/KlarnaRequestFactory.java]: it will be printed by exception catcher: https://github.com/commercetools/commercetools-payone-integration/blob/0a1ca55e4fda35ac165790beb7e610e67cd178e7/service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java#L78

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/KlarnaAuthorizationRequest.java (hunk 4 above)
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/KlarnaAuthorizationRequest.java]: it's not recommended for on-line shops, but still can be use by other suppliers/consumers (warehouses, robots, export/import jobs and so on)

File: docs/Field-Mapping.md (hunk 5 above)
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: 87278a9f9c289faee96f2324cb1b92249b7b4456

File: docs/Field-Mapping.md [hunk 13]
```diff
@@ -283,22 +283,30 @@ All payment methods:
     * `clearing_bankbic` ->  `paidFromBIC` 
    
 `INVOICE-KLARNA`:
-  * _Required_ `language` -> custom field `languageTag` of Type String on the CT Payment
-    * (may be moved to the Order, Cart and Customer Objects later)
+  * _Required_ 
+    * `language` -> custom field `languageTag` of Type String on the CT Payment
+      (may be moved to the Order, Cart and Customer Objects later)
```
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: it was outdated: 10296102729383be554c5297c5f046a37afdcc74

File: docs/Field-Mapping.md [hunk 14]
```diff
@@ -283,22 +283,30 @@ All payment methods:
...
+      
+    * `birthday` -> String custom field `birthday` in Payment in _YYYMMDD_ format.
+    * `ip` -> the IP address of the user -> String custom field `ip` in Payment.
```
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: 41df144cb334508d901da7cf3d887ab530a4ea4a

File: docs/Field-Mapping.md [hunk 15]
```diff
@@ -283,22 +283,30 @@ All payment methods:
...
 `INVOICE-KLARNA`:
-  * _Required_ `language` -> custom field `languageTag` of Type String on the CT Payment
-    * (may be moved to the Order, Cart and Customer Objects later)
+  * _Required_ 
+    * `language` -> custom field `languageTag` of Type String on the CT Payment
+      (may be moved to the Order, Cart and Customer Objects later)
+      
+    * `birthday` -> String custom field `birthday` in Payment in _YYYMMDD_ format.
+    * `ip` -> the IP address of the user -> String custom field `ip` in Payment. 
+    * `gender` -> The value is suppressed to first lowercase character -> String custom field `gender` in Payment
```
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: 940b6f45aebfbc0eb837b48fa0804f561e7db2b4

File: docs/Field-Mapping.md (hunk 6 above)
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: 73819ef6

File: docs/Field-Mapping.md [hunk 16]
```diff
@@ -307,13 +315,12 @@ All payment methods:
...
  
- * `birthdate` and `vatid` : these fields are only available on the CT Customer and not on the Cart. I.e. Guest checkouts
-   cannot do some payment methods.
-   * If the fields `dateOfBirth` (type Date) and `vatId` (type String) respectively are set as custom object on the 
-     Cart / Order and have the right type they are used and take precedence over the Fields on the Customer Object
- * `gender`: Check the Cart for a custom field `customerGender`, as fallback check the Customer for a custom field 
-    named `gender`. If the first existing is of Type `Enum` and has a value `Male` or `Female` -> use that one as `f` or `m` respectively.  
- * `ip`: Check for a custom Field `customerIPAddress` of type `String` on the CT Cart / Order. 
+ * If the fields `dateOfBirth` (type Date) and `vatId` (type String) respectively are set as custom object on the 
+     Cart / Order and have the right type they are used and take precedence over the Fields on the Customer Object
```
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: updated doc to actual behavior: 10296102729383be554c5297c5f046a37afdcc74 and 07b56df2870ff3de7f9ec4c2ae4b841387036c58

File: docs/Field-Mapping.md [hunk 17]
```diff
@@ -307,13 +315,12 @@ All payment methods:
...
- * `birthdate` and `vatid` : these fields are only available on the CT Customer and not on the Cart. I.e. Guest checkouts
-   cannot do some payment methods.
-   * If the fields `dateOfBirth` (type Date) and `vatId` (type String) respectively are set as custom object on the 
-     Cart / Order and have the right type they are used and take precedence over the Fields on the Customer Object
- * `gender`: Check the Cart for a custom field `customerGender`, as fallback check the Customer for a custom field 
-    named `gender`. If the first existing is of Type `Enum` and has a value `Male` or `Female` -> use that one as `f` or `m` respectively.  
- * `ip`: Check for a custom Field `customerIPAddress` of type `String` on the CT Cart / Order. 
+ * If the fields `dateOfBirth` (type Date) and `vatId` (type String) respectively are set as custom object on the 
+     Cart / Order and have the right type they are used and take precedence over the Fields on the Customer Object 
+ * If `birthday` is set in Payment custom field - this value overrides value from CT Customer above.
```
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: It is a _type_, so from uppercase: 07b56df2870ff3de7f9ec4c2ae4b841387036c58

File: docs/Migration-Guide.md [hunk 18]
```diff
@@ -87,4 +88,42 @@ You have to change the payment handling URL:
 
 ### 5. PROFIT
 
-Congratulations, you are done!
\ No newline at end of file
+Congratulations, you are done!
+
+## Add Klarna payment
```
Comment (andrii-kovalenko-ct) [inline docs/Migration-Guide.md]: 6d06b530e5b5448eabf6c8879946324075d00121

File: docs/Migration-Guide.md [hunk 19]
```diff
@@ -87,4 +88,42 @@ You have to change the payment handling URL:
...
+Update custom types: add IP, gender.
+Klarna library updates:
+  1. [Create Klarna account](https://klarna.com/buy-klarna/our-services/klarna-account), if not created yet.
+     For test purpose - [create test Klarna account](https://developers.klarna.com/en/de/kpm/apply-for-test-account).
+  
+  2. Activate Klarna mode in Payone. **Note**: this process may take couple of days, so do it beforehand.
+     
+     2.1 Supply Klarna API credentials (_EID_ and _Shared Secret_) to your Payone Merchant Manager.
+  
+  3. Extend Klarna payment custom type in commercetools platform. In addition to standard _languageCode_ and _reference_
```
Comment (andrii-kovalenko-ct) [inline docs/Migration-Guide.md]: 316a23dbf1dcd4479efc8e3d6a4c46e3e0edb3af

File: docs/Migration-Guide.md [hunk 20]
```diff
@@ -87,4 +88,42 @@ You have to change the payment handling URL:
...
+     fields add the following:
+
+     <table>
+     <tr><th>Name</th><th>Type</th><th>Required</th><th>Input Hint</th><th>Notes</th></tr>
+     <tr><td><b><code>ip</code></b></td><td rowspan="4"><i>String</i></td><td rowspan="4"><b>true</b></td><td rowspan="4"><i>SingleLine</i></td><td></td></tr>
+     <tr><td><b><code>gender</code></b></td><td>Only first lowercase character is used, see <a href="/blob/master/service/src/main/java/com/commercetools/pspadapter/payone/mapping/MappingUtil.java#L181-L181">MappingUtil.getGenderFromPaymentCart()</a></td></tr>
+     <tr><td><b><code>birthday</code></b></td><td>If this field is empy - the service will try to apply 
+                      <a href="http://dev.commercetools.com/http-api-projects-customers.html#customer">Customer's dateOfBirth</a>, 
+                      but this field is an optional, also it is not available for anonymouse purchases.
```
Comment (andrii-kovalenko-ct) [inline docs/Migration-Guide.md]: 7b7ca8d13afa28ac224b6300dfca4900e1ae202f

File: docs/Migration-Guide.md [hunk 21]
```diff
@@ -87,4 +88,42 @@ You have to change the payment handling URL:
...
+     </table>
+     
+  4. In the shop's frontend:
+     
+     4.1 Apply filters to the payment method, for instance, allow to select Klarna only if delivery/shipping address is
+     in the list of activated countries (by default Klarna is disabled for all countries).
+     
+     4.2 Ensure Klarna mandatory fields from the table above are populated, also note that **`telephonenumber`** is 
+     mandatory for Klarna, so ensure it is set in [`CartLike#billingAddress`](http://dev.commercetools.com/http-api-projects-carts.html#cart)
+     before handling the cart/payment.
```
Comment (andrii-kovalenko-ct) [inline docs/Migration-Guide.md]: 9e446d67e5060e1f732d8ba050218e3602d5d6e7

File: docs/Project-Lifecycle.md (hunk 7 above)
Comment (andrii-kovalenko-ct) [inline docs/Project-Lifecycle.md]: 5759f8f9d68551258062063ed7a8e47ae3992ca5

File: docs/Field-Mapping.md (hunk 5 above)
Comment (butenkor) [inline docs/Field-Mapping.md]: Link for implementation detail is ok but i think we should document also here in which order language is being deduced and which ISO standard is expected.

File: docs/Field-Mapping.md (hunk 16 above)
Comment (butenkor) [inline docs/Field-Mapping.md]: why did you remove vatIt if it is currently also supported/mapped?

File: docs/Project-Lifecycle.md (hunk 7 above)
Comment (butenkor) [inline docs/Project-Lifecycle.md]: You should setup manually only currency, language and country as prices are part of the product and can be created over API (IT -Tests). Also tax category can be created and assigned over API (test)

File: docs/Project-Lifecycle.md (hunk 8 above)
Comment (butenkor) [inline docs/Project-Lifecycle.md]: please create an issue for that and we will need it for new integration service also @floriansattler

File: service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java (hunk 9 above)
Comment (butenkor) [inline service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java]: please create an issue

File: service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java (hunk 10 above)
Comment (butenkor) [inline service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java]: please create an issue

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/PayonePostServiceImpl.java (hunk 11 above)
Comment (butenkor) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/PayonePostServiceImpl.java]: What speaks against putting data directly into map as `map.add("a[1]", 5);` instead of converting it here again?

File: docs/Field-Mapping.md (hunk 16 above)
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: that was wrong comment: `vatId` is not a part of Klarna request, but optional parameter of common Payone request. It is described in [the same document above](https://github.com/commercetools/commercetools-payone-integration/blob/10296102729383be554c5297c5f046a37afdcc74/docs/Field-Mapping.md#commercetools-payment-resource-mapping)

File: docs/Project-Lifecycle.md (hunk 7 above)
Comment (andrii-kovalenko-ct) [inline docs/Project-Lifecycle.md]: of course we could extend our integration tests to be self-sufficient and self-configured, but this requires additional task/time, which is not a part of this task

Comment (andrii-kovalenko-ct) [inline docs/Project-Lifecycle.md]: https://github.com/commercetools/commercetools-payone-integration/issues/161

File: docs/Project-Lifecycle.md (hunk 8 above)
Comment (andrii-kovalenko-ct) [inline docs/Project-Lifecycle.md]: https://github.com/commercetools/commercetools-payone-integration/issues/161

File: service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java (hunk 10 above)
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java]: https://github.com/commercetools/commercetools-payone-integration/issues/162

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/PayonePostServiceImpl.java (hunk 11 above)
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/PayonePostServiceImpl.java]: `BaseRequest` is a POJO bean, it doesn't have dynamic properties with names like `a[1]`, but it could have array/list properties instead. The [`BaseRequest#toStringMap`](https://github.com/commercetools/commercetools-payone-integration/blob/58bd98cb85012e6dd0a34b3b276590e435b92efb/service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/common/BaseRequest.java#L82-L82) can convert the bean object to map of key=value pairs, where _key_ will be a list name, _value_ - `List` itself.

Unfortunately, lists format is not specified in HTTP/W3C standards, so every service could treat them as they want. By default `Unirest.post(Map)` will represent a list as: `a=[x, y , z]` (e.g. one value with 3 elements), whereas Payone requires them in format `a[1]=x&a[2]=y&a[3]=z` (e.g. 3 values).

That's why we should convert default Map representation to Payone specific request format.

File: docs/Field-Mapping.md (hunk 5 above)
Comment (andrii-kovalenko-ct) [inline docs/Field-Mapping.md]: a89bc1ea

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/BaseKlarnaRequest.java (hunk 12 above)
Comment (butenkor) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/BaseKlarnaRequest.java]: Why to use lambda or functional approach if it is not required and actually forces to use things which are meant to be used in rather different scenarios?

Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/klarna/BaseKlarnaRequest.java]: Ok, re-factored to **true** way: b9551ff2

**Comment by andrii-kovalenko-ct** [comment]
@butenkor are you satisfied with the latest changes? i'd like to make a release

File: service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java (hunk 9 above)
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java]: https://github.com/commercetools/commercetools-payone-integration/issues/164

Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java]: https://github.com/commercetools/commercetools-payone-integration/issues/165
